File: presentation/src/main/java/dev/olog/presentation/utils/rx/SlidingUpPanelEventsObservable.java
package dev.olog.presentation.utils.rx;

import android.view.View;

import com.sothree.slidinguppanel.SlidingUpPanelLayout;

import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.android.MainThreadDisposable;

import static com.sothree.slidinguppanel.SlidingUpPanelLayout.PanelState.COLLAPSED;
import static dev.olog.presentation.utils.ThreadUtilsKt.assertMainThread;

class SlidingUpPanelEventsObservable extends Observable<SlidingPanelState> {

    private final SlidingUpPanelLayout view;

    SlidingUpPanelEventsObservable(SlidingUpPanelLayout view) {
        this.view = view;
    }

    @Override
    protected void subscribeActual(Observer observer) {
        assertMainThread();
        Listener listener = new Listener(view, observer);
        observer.onSubscribe(listener);
        view.addPanelSlideListener(listener.panelSlideListener);
    }

    final class Listener extends MainThreadDisposable {

        private final SlidingUpPanelLayout slidingUpPanelLayout;
        private final SlidingUpPanelLayout.PanelSlideListener panelSlideListener;

        Listener(SlidingUpPanelLayout slidingUpPanelLayout, Observer<SlidingPanelState> observer) {

            observer.onNext(new SlidingPanelState(COLLAPSED, COLLAPSED));

            this.slidingUpPanelLayout = slidingUpPanelLayout;
            this.panelSlideListener = new SlidingUpPanelLayout.SimplePanelSlideListener(){
                @Override
                public void onPanelStateChanged(View panel, SlidingUpPanelLayout.PanelState previousState, SlidingUpPanelLayout.PanelState newState) {
                    if (!isDisposed()){
                        observer.onNext(new SlidingPanelState(previousState, newState));
                    }
                }
            };
        }

        @Override
        protected void onDispose() {
            slidingUpPanelLayout.removePanelSlideListener(panelSlideListener);
        }
    }
}

File: presentation/src/main/java/dev/olog/presentation/utils/rx/SlidingUpPanelEventsObservable.kt
package dev.olog.presentation.utils.rx

import android.view.View
import com.sothree.slidinguppanel.SlidingUpPanelLayout
import com.sothree.slidinguppanel.SlidingUpPanelLayout.PanelState.COLLAPSED
import dev.olog.shared_android.assertMainThread
import io.reactivex.Observable
import io.reactivex.Observer
import io.reactivex.android.MainThreadDisposable

internal class SlidingUpPanelEventsObservable(
        private val view: SlidingUpPanelLayout

) : Observable<SlidingPanelState>() {

    override fun subscribeActual(observer: Observer<in SlidingPanelState>) {
        assertMainThread()
        val listener = Listener(view, observer)
        observer.onSubscribe(listener)
        view.addPanelSlideListener(listener.panelSlideListener)
    }

    internal inner class Listener(
            private val slidingUpPanelLayout: SlidingUpPanelLayout,
            observer: Observer<in SlidingPanelState>

    ) : MainThreadDisposable() {

        val panelSlideListener = object : SlidingUpPanelLayout.SimplePanelSlideListener() {
            override fun onPanelStateChanged(panel: View?, previousState: SlidingUpPanelLayout.PanelState?, newState: SlidingUpPanelLayout.PanelState?) {
                if (!isDisposed) {
                    observer.onNext(SlidingPanelState(previousState!!, newState!!))
                }
            }
        }

        init {
            observer.onNext(SlidingPanelState(COLLAPSED, COLLAPSED))
        }

        override fun onDispose() {
            slidingUpPanelLayout.removePanelSlideListener(panelSlideListener)
        }
    }
}
